### ISSUE TITLE
Single file components

### ISSUE THREAD

**Issue by vb8448** [EmilStenstrom/django-components#183]
Have you ever thought to implement a single file component, similar to vuejs ?

Eg.:
```python
@component.register("calendar")
class Calendar(component.Component):
    # This component takes one parameter, a date string to show in the template
    def get_context_data(self, date):
        return {
            "date": date,
        }
   
    template = """
        <div class="calendar-component">Today's date is <span>{{ date }}</span></div>
    """
    css = """
        .calendar-component { width: 200px; background: pink; }
        .calendar-component span { font-weight: bold; }
    """
    js = """
        (function(){
            if (document.querySelector(".calendar-component")) {
                document.querySelector(".calendar-component").onclick = function(){ alert("Clicked calendar!"); };
            }
        })()
    """
```

**Comment by vb8448** [comment]
> you lose all syntax highlighting in editors. Is there a way to get syntax highlighting some other way?

Pycharm seems to support [language injections](https://www.jetbrains.com/help/pycharm/using-language-injections.html#inject_language) and I guess also vscode does. Probably a specific extension will be needed.

**Comment by EmilStenstrom** [comment]
I did some reading on this, and found a couple of crazy solutions such as https://github.com/twidi/mixt/ which have solved this in a mix of crazy and fantastic. But there's really nothing good out there. Language injections works in pycharm, but not in vscode. This also bypasses Django's template loading system, which loses caching. 

Overall, I think using strings like in your quick example would quickly break down for bigger components. So right now it's a "wontfix".

I do really like to have a solution for this, but I just don't see a solution I'd like to recommend to users. Losing syntax highlighting in editors seems like the major one right now.

**Comment by aymericderbois** [comment]
I love the idea of having the template directly in a component variable. This is especially useful when you want to make small components!
Unfortunately for syntax highlighting, I think there is no solution. Even with Pycharm, the support of language injections is limited.

Nevertheless, as this is the kind of functionality I could use, I tried to make some tests and I made a quick development that allows to have "inline template".

Here is a "countdown" component, with the template directly in the component : 
```python
@component.register("countdown")
class Countdown(component.Component):
    template_string = """
        <span
          data-ended-at-timestamp="{{ until.timestamp }}"
          x-data="{
                interval: null,
                init() {
                    this.endTimestamp = parseInt($el.dataset.endedAtTimestamp);
                    this.processCountdown();
                    this.interval = setInterval(() => this.processCountdown(), 1000);
                },
                processCountdown() {
                    const currentTimestamp = Math.floor(Date.now() / 1000)
                    const remainingTime = this.endTimestamp - currentTimestamp;
                    if ($refs.days === undefined && this.interval !== null) {
                        clearInterval(this.interval)
                    } else {
                        $refs.days.innerHTML = Math.floor(remainingTime / 60 / 60 / 24)
                        $refs.hours.innerHTML = Math.floor(remainingTime / 60 / 60) % 24
                        $refs.minutes.innerHTML = Math.floor(remainingTime / 60) % 60
                        $refs.seconds.innerHTML = Math.floor(remainingTime) % 60
                    }
                }
            }"
        >
            <span x-ref="days"></span>j
            <span x-ref="hours"></span>h
            <span x-ref="minutes"></span>ms
            <span x-ref="seconds"></span>s
        </span>
    """

    def get_context_data(self, until: datetime):
        return {
            "until": until,
        }
```

The commit applying the modification is here : https://github.com/aymericderbois/django-components/commit/d650a9250812e708f441f4c4b488748a402f8a6c

The advantage of this implementation is to keep the use of the Django template system. I did this quickly, so I guess it could be greatly improved.

**Comment by EmilStenstrom** [comment]
Thanks for taking the time to show how it could look like. 

I **really** don't like the developer ergonomics of having to write html, css, and js inside of python strings. If there was (when there is?) a way to enable syntax highlighting in some clever way, I would be OK with this change. But as it stands, I don't think we should encourage that kind of code, and instead only support having separate files in a directory. 

Since all the changes are in the Component class, I think you could make your own component base, and inherit from that inside your project? That way, we wouldn't have to

**Comment by EmilStenstrom** [comment]
Hmm. I keep thinking about how to get this done, because it would a lot simpler. Maybe we should write [a custom VSCode extension](https://code.visualstudio.com/api) that can highlight our custom file format?

**Comment by EmilStenstrom** [comment]
I have opened the following PR to try to get editors to highlight this code: https://github.com/microsoft/vscode/issues/169560

**Comment by EmilStenstrom** [comment]
The Python extension people with VSCode are currently thinking about adding support for this (!). All we need right now are more upvotes on this ticket: https://www.github.com/microsoft/pylance-release/issues/3874

Is this something that you, or your colleague, or maybe your grandmother would find useful? Please add a 👍 to that issue, and tell your friends! We're doing this!!

**Comment by rtr1** [comment]
I was able to get syntax highlighting to work using [https://github.com/samwillis/python-inline-source](python-inline-source)

![image](https://github.com/EmilStenstrom/django-components/assets/8939937/6dcfdada-2bb4-4d2c-80a4-762d345cb399)

**Comment by rtr1** [comment]
I got two Tailwind VS Code extensions to work inside Python strings.

- [Tailwind CSS IntelliSense](https://marketplace.visualstudio.com/items?itemName=bradlc.vscode-tailwindcss&ssr=false#review-details) (for Tailwind autocompletion)
- [Inline Fold](https://marketplace.visualstudio.com/items?itemName=moalamri.inline-fold&ssr=false#review-details) (to hide the long Tailwind lines)

Now with `Tailwind`, [Alpine](https://alpinejs.dev/), and [HTMX](https://htmx.org/), I'm making Django components with only the `.py` file. I love it!

In VS Code's `settings.json`:
```json
     "editor.quickSuggestions": {
        "strings": true
     },
     "tailwindCSS.includeLanguages": {
        "python": "html"
      },
      "[py]": {
        "tailwindCSS.experimental.classRegex": [
          "\\bclass=\"([^\"]*)\"",
        ]
    },
    "inlineFold.supportedLanguages": [
        "python"
    ]
```
And then using [Tailwind CLI](https://tailwindcss.com/blog/standalone-cli), in `tailwind.config.js`:
```javascript
module.exports = {
  content: [
    './apps/**/templates/**/*.html',
    './components/**/*.html',
    './components/**/*.py'
  ],
  theme: {
    extend: {},
  },
  plugins: [],
}
```

**Comment by EmilStenstrom** [comment]
@rtr1 Sounds great! Do you run a fork of django-components that supports inline strings? Would you mind submitting a PR?

**Comment by EmilStenstrom** [comment]
Ah, then I understand. Mind posting an example of how a real component looks with this way of writing it?

**Comment by rtr1** [comment]
Here are some examples with some sample data.

A `title` component that sits inside a `header` component, and a `location_menu` component that's a dropdown `select` menu.

I think it should be possible to create a generic component library with this approach, like I'm working on a multi-select dropdown that's similar to the `location_menu` except it's using Alpine for the added functionality.

```python
# title.py
from django_components import component
from django.template import Template
from sourcetypes import html

@component.register("title")
class Title (component.Component):
    template_name = ...

    def get_context_data(self, text): return {'text': text}
    
    def get_template(self, context, template_name: str | None = None):
        template: html = """<div class="ml-2"><h1>{{text}}</h1></div>"""
        return Template(template)
```

```python
# header.py
from django.template import Template
from django_components import component
from sourcetypes import html

template: html = """
    {% load component_tags %}
    <nav class="flex items-center justify-between flex-wrap bg-teal-500 p-5 text-white font-semibold text-xl tracking-tight">
        {% component "title" text="Company Report" %}
    </nav>
"""

@component.register("header")
class Header (component.Component):
    template_name = ...
    
    def get_template(self, *args, **kwargs):
        return Template(template)
```

```python
# location_menu.py
from django.template import Template
from django_components import component
from sourcetypes import html
from apps.company.models import Location

template: html = """
<div class="flex flex-col">

    <label for="location" class="block text-gray-600 text-sm font-bold m-2">Location</label>

    <select name="location" id="location" class="block w-full px-4 py-2 mb-1 border border-gray-300 bg-white rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:border-indigo-500 text-sm">
        <option selected>Choose a location</option>
        {% for location in locations %}
            <option value="{{location.id}}">{{location.name}}</option>
        {% endfor %}

    </select>

</div>
"""

@component.register("location_menu")
class LocationMenu (component.Component):
    template_name = ...

    def get_context_data(self):
        return { 'locations': Location.objects.all().order_by("name") }
    
    def get_template(self, *args, **kwargs):
        return Template(template)
```

You can see the Tailwind code looks ugly when viewing the plain-text above, but in VS Code the components look pretty nice with Tailwind details collapsed.

![image](https://github.com/EmilStenstrom/django-components/assets/8939937/d36b58c6-0840-4ea7-bf95-eabda05f05ed)

And auto-completion works even inside the Python strings:

![image](https://github.com/EmilStenstrom/django-components/assets/8939937/469fce84-06b2-43d3-9b0a-dfe751530fcc)

And it feels very much like working with React/Vue/etc, being able to build very small components and piece them together.

![image](https://github.com/EmilStenstrom/django-components/assets/8939937/21ad08d4-f340-496a-8fb2-ce9fe7399eb8)

**Comment by rtr1** [comment]
@EmilStenstrom is there currently a way to present a component as a view?

For instance, is it possible to pass the component to `urlpatterns`, perhaps something similar to how class-based views have the `.as_view()` method, or some concept like `DefaultRouter` from DRF that lets us map endpoints to components?

**Comment by EmilStenstrom** [comment]
@rtr1 Thank you for that example above, really inspiring. I think we should work on allowing single file components based on python-inline-source soon.

About "component as view": It's not possible to do this today, feel free to open a separate issue to track this. I think this is a really interesting track to explore, with the end goal to be able to build something like Phoenix LiveView in Django... It's not clear to me wether this should be part of this library or a separate one that depends on this one...

**Comment by nerdoc** [comment]
I dislike the idea of single file components, not because it is not convenient, but it has major drawbacks for devs. Even when syntax highlighting works, you loose things like auto suggestions, click-to-navigate to code, IDE's understanding of the code structure etc.
There is another framework: [Tetra](tetraframework.com), that solves this using the VSCode plugin. PyCharm does not works at all yet besides Syntax highlighting (manually). So -1 for this, efforts are better spent into other features IMHO...

**Comment by rtr1** [comment]
I see at least two distinct scenarios.

1. One is what @nerdoc brings up. I think if one has extensive JS or CSS in a component, this a very valid point, and becomes more valid as the complexity of those pieces of the puzzle increase.

2. Two is using a “low/no JS” approach, like utilizing Tailwind and Alpine. This scenario can basically already be done with existing tooling, IDE extensions, etc.

So it may be the case that, the only practical application of SFC is one which is already possible, and no further changes would be needed. At least, it’s worth considering if this is the case or not.

**Comment by EmilStenstrom** [comment]
@nerdoc To be clear: The suggestion here is not to replace all existing multi-file components with single-file components. but to add single-file components as an option. If you have a complex component, that would likely be in multiple files, but simple ones could be in a single file. Again, this would be optional for you as a user, and I wouldn't want to break backwards compatibility by removing multi-file components.

**Comment by nerdoc** [comment]
Yeah, thats what I thought. As long as the "traditional" approach stays, a single file approach would reallly be a perfect addition.

**Comment by dylanjcastillo** [comment]
Would it be too crazy to have something in between? Not exactly a SFC, more like a single-file UI component + python component:

Something like this:

UI component:
```html
<style>
    ...
</style>
<template>
    <div id="table-view-inner">
        <div class="flex flex-row justify-end">
            <p>
                Showing <span class="font-medium">{{ table_data|length }}</span> results
            </p>
        </div>
        <form class="flex flex-row align-baseline mb-4"
              hx-get="{% url 'table_view' %}"
              hx-target="#table-view-inner"
              hx-trigger="submit, change from:select[name='search_sorting']">
            ....
            <select id="search_sorting"
                    name="search_sorting"
                    class="w-1/5 bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500">
                <option value="most-recent"
                        {% if widget_state.search_sorting == 'most-recent' %}selected{% endif %}>Most recent</option>
                <option value="oldest"
                        {% if widget_state.search_sorting == 'oldest' %}selected{% endif %}>Oldest</option>
                <option value="most-cited"
                        {% if widget_state.search_sorting == 'most-cited' %}selected{% endif %}>Most cited</option>
                <option value="least-cited"
                        {% if widget_state.search_sorting == 'least-cited' %}selected{% endif %}>Least cited</option>
            </select>
        </form>
        {% for journal in table_data %}
            {% component 'table_row' journal=journal %}
        {% endfor %}
    </div>
</template>
<script>
    (function() {
        document.addEventListener("htmx:configRequest", function(event) {
            console.log(event.detail.triggeringEvent);
            if (event.detail.triggeringEvent.submitter) {
                event.detail.parameters["event_source"] =
                    event.detail.triggeringEvent.submitter.id;
            } else {
                event.detail.parameters["event_source"] =
                    event.detail.triggeringEvent.target.id;
            }
        });
    })();
</script>
```

Python Component:
```python
from typing import Any, Dict
from django_components import component


@component.register("table_view")
class Calendar(component.Component, use_ui_component=True):
    template_name = "table_view/table_view.html"

    def get_context_data(self, table_data, widget_state) -> Dict[str, Any]:
        return {
            "table_data": table_data,
            "widget_state": widget_state,
        }
```

The UI component would be an HTML file, so you get syntax highlighting, etc for free but you also get more LOB on the UI side of things.

**Comment by EmilStenstrom** [comment]
@dylanjcastillo With HTMX and Tailwind, you are not as likely to need any style and js files at all. So I guess you can do that already...

I think the value here is having the python file contain everything you need. If it's two or four files matters less. The VSCode extension above solves inlines color highlighting nicesly.

So to me, this is just a matter of who will write the PR?

**Comment by dylanjcastillo** [comment]
Hey @EmilStenstrom, I agree that the single Python file is ideal, but you'd lose things like auto-formatting of the file (e.g., Prettier, djlint), which would reduce this to only very simple components.

OTOH, I'd like to get a bit more understanding of the codebase, and this could be a good starting point. So after I finish a project by the end of the month, I'll take a shot at this.

**Comment by ekaj2** [comment]
1) Neovim has this plugin which works great for this concept: https://github.com/AndrewRadev/inline_edit.vim/
2) I'd much rather have a single component in an html file than a python file as including <script> tags is so common that most editors LSP + highlighting support it out of the box. The python bits could be inferred based on filename, etc. With HTMX and Tailwind, I almost always just copy/paste the boilerplate, rename the class + component to match my file name.

**Comment by EmilStenstrom** [comment]
@ekaj2 Thanks for the neovim link, lots of good inspiration!

As I've said before, my view is that it's already possible to embed script and css in the template file if you want. Just inline the CSS with tailwind and put the js in a script tag. The python file is then likely small enough to be easy to do. I've created an issue to maybe make this workflow even easier #337, but I'd say it's a nice to have.

@all Single file components in the python file is likely for small components, with just a few lines of template code, or CSS, which in my experience most components are. This means that automatic formatting within those strings is not as useful, although if someone finds a way to support that it would of course be nice.

So let's focus on putting everything in the python file for in this issue.

# Design suggestion
There are four languages that we want to support in this single file: Python, HTML, CSS, and JS.

Python: Since we already are within a python file, that case is solved :) ✅ 
HTML: Importing Template from django.template as in https://github.com/EmilStenstrom/django-components/issues/183#issuecomment-1553388161 solves this nicely I think. No need for a separate template file.
**CSS and JS: This is what I'd like to fix, supporting putting your code in strings instead of as links to separate files.**

Syntax idea:

```python
@component.register("location_menu")
class LocationMenu(component.Component):
    def get_context_data(self):
        return {'locations': Location.objects.all().order_by("name")}
    
    def get_template(self, *args, **kwargs):
        return Template("""
            <ul class="menu">
                {% for location in locations %}
                    <li><a href="/navigate/{{ location.id }}/">{{ location.name }}</a></li>
                {% endfor %}
            </ul>
        """)

    def get_css(self, *args, **kwargs):
        return """
            .menu {
                display: inline-block;
                width: 100px;
            }
        """

    def get_js(self, *args, **kwargs):
        return """
            window.addEventListener("load", (event) => {
                console.log("Page loaded!");
            });
        """
```

And maybe there should be an even simpler API where you just put your code in a variable with a special name:

```python
@component.register("location_menu")
class LocationMenu(component.Component):
    def get_context_data(self):
        return {'locations': Location.objects.all().order_by("name")}
    
    template = """
        <ul class="menu">
            {% for location in locations %}
                <li><a href="/navigate/{{ location.id }}/">{{ location.name }}</a></li>
            {% endfor %}
        </ul>
    """)
    
    css = """
        .menu {
            display: inline-block;
            width: 100px;
        }
    """
    
    js = """
        window.addEventListener("load", (event) => {
            console.log("Page loaded!");
        });
    """
```

**Comment by EmilStenstrom** [comment]
Fixed by https://github.com/EmilStenstrom/django-components/pull/361, thanks @dylanjcastillo!

**Comment by EmilStenstrom** [comment]
This is now released as version 0.32 on PyPI: https://pypi.org/project/django-components/0.32/
